Issue 40415: Server JavaScript Console not displaying trigger messages fired by ETL

labkey-martyp opened the pull request:
#### Rationale
Update SessionAppender to store appender infos for cross thread access.  This is to allow trigger script console messages triggered from an ETL pipeline job to use the same appender info as the current view context.

#### Related Pull Requests
https://github.com/LabKey/dataintegration/pull/46

#### Changes
Add register mechanism to save appender infos to concurrent map.  Add functions to get appender info by key and to set the current thread's appender info to an already existing appender info.

labkey-martyp committed:
Register Session AppenderInfo for use in other threads

- Allow pipeline job to access session appender of main thread to print trigger script messages to server javascript console
- Register appenderInfos in concurrent hashmap and allow retrieval by key

labkey-martyp committed:
Update comment

labkey-martyp committed:
Update function name

labkey-martyp committed:
Update comment

labkey-matthewb commented on api/src/org/labkey/api/util/SessionAppender.java:
```diff
@@ -45,7 +48,25 @@
     }
 
     private static final ThreadLocal<AppenderInfo> localInfo = new ThreadLocal<>();
+    private static final Map<String, AppenderInfo> appenderInfos = new ConcurrentCaseInsensitiveSortedMap<>();
```
Roaches check in, but they don't check out!

This needs to use some sort of WeakHashMap, or a SessionListener to remove AppendInfo that are associated with sessions that have gone away.

labkey-martyp replied:
Dang, yes forgot the cleanup on this.  I went with a ConcurrentReferenceMap which is essentially a concurrent WeakHashMap.

labkey-martyp committed:
Code review feedback

labkey-matthewb commented on api/src/org/labkey/api/util/SessionAppender.java:
```diff
@@ -48,16 +48,16 @@
 
     private static final ThreadLocal<AppenderInfo> localInfo = new ThreadLocal<>();
-    private static final Map<String, AppenderInfo> appenderInfos = new ConcurrentCaseInsensitiveSortedMap<>();
+    private static final Map<String, AppenderInfo> appenderInfos = new ConcurrentReferenceHashMap<>();
```
What a great class.  You should probably specify ReferenceType.WEAK, instead of the default ReferenceType.SOFT.

labkey-matthewb replied:
Also, probably worth a comment explaining how/when the cleanup should work.  e.g. References to the SessionAppender should go away when the session is destroyed allowing the ConcurrentReferenceHashMap to release the entry.

labkey-martyp replied:
Oh yeah, I missed that was a soft reference by default. Specified to WEAK reference and added a comment about garbage collection on it.

labkey-martyp committed:
Update to use weak reference

- Specify ConcurrentReferenceHashMap uses WEAK reference instead of SOFT which is default.
- Add comment explaining cleanup

labkey-martyp committed:
Merge branch 'develop' into fb_40415_server_javascript_console
